### ISSUE TITLE
Added constraints on ParameterInformation.label preclude interesting use cases

### ISSUE THREAD

**Issue by oblitum** [Microsoft/language-server-protocol#640]
Beyond adding support for inclusive start and exclusive end offsets to spec, commit 3854f1c added additional constraints to `ParameterInformation.label` when in string form, which can preclude neat functionality for users:

> *Note*: A label of type string must be a substring of its containing signature label.

For example, before, compliant client and servers could take advantage of not having such limitation to provide terser prototype form on `SignatureInformation.label`, while providing full parameter types on each `ParameterInformation.label`. Types can be rather long, so having full form just for active parameter that clients can pick from `parameters`, while having short form of the rest (from `SignatureInformation.label`), can provide ideal UX.

### Proposal

To solve this, while still possibly preserve current constraints, I think the best protocol would not provide `label: string | [number, number]` but instead a `label: string | [number, number] | [[number, number], string]`:

```ts
label: string                // substring constrained to SignatureInformation.label
     | [number, number]      // offsets constrained to SignatureInformation.label
     | [                     // (tuple, not array)
           [number, number], // offsets constrained to SignatureInformation.label
           string            // unconstrained "full" ParameterInformation.label
       ]
```

**Comment by oblitum** [comment]
It can also be terser if the recently added ` | [number, number]` is dropped in favor of the new one:

```ts
label: string                // substring constrained to SignatureInformation.label
     | [                     // (tuple, not array)
           [number, number], // offsets constrained to SignatureInformation.label
           string            // unconstrained "full" ParameterInformation.label
       ]
```

**Comment by dbaeumer** [comment]
@oblitum I see your point however what would the client have done with that information. Can you give me an example about what a client rendered in the case you described. A client is not able to reconstruct a signature from the parameter info since it usually doesn't know what parameter separators to use.

**Comment by oblitum** [comment]
@dbaeumer I've given an example in the image. With both offsets over short signatures and detailed parameters client can do that.

**Comment by oblitum** [comment]
This gif shows how signatures can be rather noisy and how the above gives room to fix this:

![simplescreenrecorder](https://user-images.githubusercontent.com/1269815/50162825-59824300-02c6-11e9-80ce-cd178efedd8a.gif)

**Comment by oblitum** [comment]
> I am still unsure I understand this since I am not sure what you propose a client should render. Do you propose that clients show both the short and the long form?

Yes and No. On the examples above the client will have _opportunity_ to chose to show both, and this can be at user's option, if it wishes to see detailed parameter info (this due to detailed per parameter info), client is able to provide a setting for that, and show two menus or one with composed result or some other idea, otherwise user could simple be content with the short prototype with (short) current parameter highlighted (this due to the offset information).

> The current protocol already allows showing detailed information for the active parameter since clients are supposed to call signature help after each key stroke when in this mode. So the label of the `SignatureInformation` can change with every key stroke. So in your example when on the second parameter the label can be:
> 
> `foo(T<…>, U<X<A, B>, Y<C, D>>, V<…>) → W<…>`

Yes, but then this form will have to be hardcoded server side and it is unable to provide the former suggestion:

```
 U<X<A, B>, Y<C, D>>               // Current parameter in full form
   foo(t, ▁                        // User typing
foo(T<…>, *U<…>*, V<…>) → W<…>     // Cleaned up signature with short current parameter highlighted
```

> The information in the `ParameterInformation` is only used to help the client highlight certain things in the `SignatureInformation` label. In the above case the `U<X<A, B>, Y<C, D>>` part in the signature label.

The proposal idea is remove the constraint that `ParameterInformation` can't provide more detailed type than what comes in the signature.

**Comment by dbaeumer** [comment]
Point taken. However I would like to see clients using this information (or at least having a UI for this) before we start adding this to the protocol. The original intention of the label of the `ParameterInformation` was not that it could be rendered start alone in the UI.

**Comment by dbaeumer** [comment]
I am not saying it is a strange idea and I am not against such a feature. All I am saying is that the field was used different by clients so far and we tried to specify this better.

This is why I marked it as a feature request after understanding your idea.

**Comment by oblitum** [comment]
> This is why I marked it as a feature request after understanding your idea.

In this view, it's not solely feature request, since it's removing something that has never been disallowed. So much so that the original string `ParameterInfo.label` application seemed to fit better for displaying parameter type than highlighting. At `int foo(int, int, int)`, all `ParameterInfo.label` are just `int`, so that it would imply it was to be viewed as substring of signature for highlighting, instead of just parameter information, seems a stretch.

**Comment by dbaeumer** [comment]
The LSP spec is not perfect and has holes. If we detect an under specified part we usually tighten the spec even if it breaks some unintended use cases. However we are open to add these use cases back if they do make sense in the intention of the spec.

Usually a signature help info include the argument names. In your example `foo(x: int, y: int)`. The parameter info label was then `x: int` and `y:int` which works in a lot of cases but not in all. This is why we added the [number, number] option and specified what the original intend to the label was. 

I changed the sentence to this to make this more clear:

> *Note*: a label of type string should be a substring of its containing signature label. Its intended use case is to highlight the parameter label part in the `SignatureInformation.label`.

**Comment by oblitum** [comment]
OK. I don't mean to offend, but this change breaks the previous possible user interpretations of this field badly, in my opinion it has ended up misnamed now, `ParameterInformation` conveyed more generality than the actual niche application, which could simply be named `ParameterHighlighting`.

**Comment by dbaeumer** [comment]
I am open for renaming it when we add additional use cases for this. 

As said earlier if the spec is underspecified we usually try to tighten it not to add additional unintended use cases. The current change is to clarify this. Do you have a better wording to do so?

**Comment by oblitum** [comment]
Better wording for what `ParameterInformation` is able to convey now is `ParameterHighlighting`, because it's the sole job it can do, even when it comes as string. The intention of the `label` as a string would be completely clear, it's just a _highlighting_ over the `SignatureInformation.label` and nothing else. Such change is more dramatic, and I doubt it would be done, but correctly fits the actual intention that's being applied. It's more dramatic than what this proposal tries to do, which is instead to have room for the previous possible applications while still serving the niche intention without any drawback.

**Comment by oblitum** [comment]
Another approach would be to rename ParameterInformation.label as ParameterInformation.highlight.

**Comment by oblitum** [comment]
Actually,  you made realize of the `label` name itself. Better than my initial proposed would be to leave `label` as a unconstrained string, and add `highlight` as a offset range OR constrained string. This would make stuff completely clarified and cover all cases.

**Comment by oblitum** [comment]
FWIW, I'm modifying the proposal for that, just to register it correctly.

**Comment by dbaeumer** [comment]
This is a breaking change which I usually try to avoid. My current goal is to clarify the current intended behavior around signature help requests and to support cases were the parameter label is not unique without breaking any existing clients or servers (this is why I changed must to should :-)). 

If we want to support other use cases we need to think about whether we can fit it into the current request or whether we need to think about a new request clients and servers can opt into.

**Comment by dbaeumer** [comment]
Ignore my last comment. I misread your latest proposal which is not breaking. 

However my primary goal is still to clarifying the current intended behavior.

**Comment by oblitum** [comment]
> However my primary goal is still to clarifying the current intended behavior.

That's the issue, the (now rewritten) proposal is clearer at that than what has been done. "label" generally conveys a label, but now, in this specific place it's meant to mean a range over another piece of text. Change while it's early! :)

**Comment by dbaeumer** [comment]
I totally agree naming this label was a bad choice and highlight would have been the better name. But if not absolutely necessary we try to clarify things and not change intention or to break.

I am absolutely open to extend this to allow unconstrained labels that can be presented in the UI as well.

**Comment by oblitum** [comment]
I could suggest `fullLabel` in addition to the `label` that acts as a range, but it sound very convoluted, so despite attained the functionality, such addition doesn't fit a protocol in early design, I'd rather break the few clients using the `label` as `[number, number]` that was added 3 days back than causing it to look worse.

**Comment by dbaeumer** [comment]
Even if we split this up the label would be specified to be a sub label to highlight a portion in the signature label. To make it clearer that its intended use case is being such a highlight I thought it is better to have one property instead of two.

**Comment by oblitum** [comment]
I think it's revolving back to original problems. Anyway, thanks for hearing my thoughts, ¯\\\_(ツ)\_/¯.

**Comment by oblitum** [comment]
Note: Added to proposal additional ways 3854f1c is a breaking change.

**Comment by oblitum** [comment]
Added Go language example.

**Comment by MaskRay** [comment]
@oblitum's proposal looks good to me. `highlight` can specify the range of the simplified label while `label` specifies the elaborated one. There can be a user preference item to prefer one over the other and the two views can be toggled with some shortcut.

```
label: string,                 // unconstrained label
highlight?: [number, number],  // "highlight" over SignatureInformation.label
```

>  I'd rather break the few clients using the label as [number, number] that was added 7 days back than causing it to look worse.

As a server and client developer who has also watched many other implementations, I'd say yes. Breaking sooner than later just causes less pain. Really few people have noticed the Dec 13 change given that there is still a typo describing the newly added `textDocument/declaration`.

> Add support for textDocument/definition request.

> label: string | [number, number];

Another thing worth clarifying is do the offsets specify Unicode code points, or UTF-16 code units? This may hardly matter in practice, though. (I hope it cares APL, C++, Haskell and some others)

> backlog

As a non-native English speaker this intimidates me :( I hope this issue can be addressed soon. The vey few client-side developers who have observed this feature are still sitting on the fence.

**Comment by dbaeumer** [comment]
@MaskRay thanks for pointing out the wrong mentioning in the change log. I fixed that.

Just to clarify my thinking here:

- I am not against having this feature. Not at all. This is why it is still open :-)
- my first priority is to clarify the existing indented behavior before adding new once.
- if possible I try to not break clients nor servers.

If we add an un-constraint label and no client does something with it then we don't win anything. So my proposal would be to have a client capability `unconstraint` on the `parameterInformation` capability. If set servers know that they can set an unconstraint label and the the client does something meaningful with it. If not set the client will not present it. 

Having these capability flags is how we moved the protocol forward so far and I would like to keep it that way.

I will clarify the offset thing. It is the same definition as with all positions and ranges. Thanks for pointing this out.

**Comment by oblitum** [comment]
> If we add an un-constraint label and no client does something with it then we don't win anything

I don't understand this since I state in the proposal:

> In this proposal `label` just means to be a textual label, as it could have always been interpreted before.

3854f1c is the first place I see it show up in protocol. In my view unknown client is being broken with that change already. Which I'm begging to be reverted for something that's actually non-breaking.

**Comment by oblitum** [comment]
I'll declare what for me this looks like. The protocol was general on that (as stated repeatedly, the word "label" **and** the type string is not in favor of the field to be constrained/viewed as highlighting), Microsoft started to have that interpretation in their tools, then thought to add it without community feedback. The lack of RFC process for a protocol makes me very sad about language server protocol.

**Comment by oblitum** [comment]
As [I've demonstrated previously](https://github.com/Microsoft/language-server-protocol/issues/640#issuecomment-448753130), I'm here more for fixing than adding a feature I wish. If it's to insert `unconstraint` into the already sorry state, I'm not in favor, my opinion is to just leave as it is, other use-cases can go look for another protocol.

**Comment by dbaeumer** [comment]
@oblitum I can see why you think that way but this is not what is happening. The LSP protocol started as a VS Code only protocol and then the community saw the usefulness of it and ask us to move it to a more public space. Due to this some parts are under specified or with the knowledge of today would have been better specified differently. 

This being said the first goal if things are under specified is to make clear what the original intention was. Reason is that we want to make sure new clients implement this behavior and servers can rely on a certain behavior. In this specific use case it is about being able to highlight the active parameter in the signature. This got clarified by the comment to make sure clients and server understand the original intend. Since a string is a bad idea and is not sufficient we added the [number, number] however it is control by a flag. A server is not allowed to use [number, number] unless a client allows it. Based on your comments (to not break clients that use the label differently) I made the containment a should and not a must (commit: 335dd85172ca34fef1abd428b7de6048069ec69b) but made it even clearer what the intention is.

I used this approach for all specification clarification so far and as said I try to balance the different aspects as good as I can. Believe me if I could start this in a green field I would do it differently.

**Comment by oblitum** [comment]
> This being said the first goal if things are under specified is to make clear what the original intention was.

This is statement alone give me chills. It translates to me as: unknown changes can be added anytime (we never know when a new under specified thing with unknown original intentions will come up), outsiders can't do nothing against that.

**Comment by oblitum** [comment]
> Another thing worth clarifying is do the offsets specify Unicode code points,
> or UTF-16 code units? This may hardly matter in practice, though. (I hope it
> cares APL, C++, Haskell and some others)

Beyond the Unicode and Emoji those programming languages allow, couldn't LSP be minimally usable for natural languages too? Does it have to make life difficult for clients/servers that wish to provide not only programming language completion/**highlighting** but also simple textual completion/**highlighting** unified in a single protocol, without having to implement this part in another completely different kind of language server just because LSP uses `UTF-16` offsets? Going further, what about syntax highlighting, what about **code comments** in Japanese (javadoc, markdown, ...)?

@MaskRay thanks for your input.

related #376
